### PR TITLE
GAP-2537: add logic to handle mandatory questions for internal removed application forms

### PR DESCRIPTION
**Pull request by a-lor-cab**
[GAP-2537](https://technologyprogramme.atlassian.net/browse/GAP-2537)
Updates the logic to handle the Removed status of application form in the mandatory question journey

Summary of the changes and the related issue. List any dependencies that are required for this change:

## Type of change

Please check the relevant options.

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [ ] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

[GAP-2537]: https://technologyprogramme.atlassian.net/browse/GAP-2537?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 1]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "Check if grant scheme is for internal or external application",
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = Boolean.class))),
+            @ApiResponse(responseCode = "404", description = "No Grant Scheme found",
+                    content = @Content(mediaType = "application/json")),
+    })
+    public ResponseEntity<SchemeMandatoryQuestionApplicationFormInfosDto> schemeHasInternalApplication(@PathVariable Integer grantSchemeId){
+        log.info("Checking if scheme with id {} has internal application form and if it is published",grantSchemeId);
+
+        log.info("Getting scheme with id {}", grantSchemeId);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:73]: There are quite a lot of logs here, not sure we want to flood the logs like this.

Comment (a-lor-cab) [reply src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:73]: shall i move some of them to a different level (debug) or just remove them?

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 2]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+        final GrantScheme grantScheme = grantSchemeService.getSchemeById(grantSchemeId);
+        log.info("Scheme with id {} found", grantSchemeId);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:75]:
```suggestion
        log.debug("Scheme with id {} found. Getting associated advert", grantSchemeId);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 3]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+
+        log.info("Getting Advert associated to scheme with id {}", grantSchemeId);
+        final GrantAdvert advert = grantAdvertService.getAdvertBySchemeId(grantSchemeId.toString());
+        log.info("Advert with id {} found for scheme with id {}", advert.getId(), grantSchemeId);
+
+        final String webpageUrl = grantAdvertService.getExternalSubmissionUrl(advert);
+
+        log.info("Checking that the advert has link to internal or external application");
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:83]:
```suggestion
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 4]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+                    content = @Content(mediaType = "application/json")),
+    })
+    public ResponseEntity<SchemeMandatoryQuestionApplicationFormInfosDto> schemeHasInternalApplication(@PathVariable Integer grantSchemeId){
+        log.info("Checking if scheme with id {} has internal application form and if it is published",grantSchemeId);
+
+        log.info("Getting scheme with id {}", grantSchemeId);
+        final GrantScheme grantScheme = grantSchemeService.getSchemeById(grantSchemeId);
+        log.info("Scheme with id {} found", grantSchemeId);
+
+        log.info("Getting Advert associated to scheme with id {}", grantSchemeId);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:77]:
```suggestion
        log.debug("Getting Advert associated to scheme with id {}", grantSchemeId);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 5]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+        final GrantAdvert advert = grantAdvertService.getAdvertBySchemeId(grantSchemeId.toString());
+        log.info("Advert with id {} found for scheme with id {}", advert.getId(), grantSchemeId);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:79]:
```suggestion
        log.debug("Advert with id {} found for scheme with id {}", advert.getId(), grantSchemeId);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java (hunk 1 above)
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:73]:
```suggestion
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 6]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
     }
+    @GetMapping("/{grantSchemeId}/hasInternalApplication")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "Check if grant scheme is for internal or external application",
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = Boolean.class))),
+            @ApiResponse(responseCode = "404", description = "No Grant Scheme found",
+                    content = @Content(mediaType = "application/json")),
+    })
+    public ResponseEntity<SchemeMandatoryQuestionApplicationFormInfosDto> schemeHasInternalApplication(@PathVariable Integer grantSchemeId){
+        log.info("Checking if scheme with id {} has internal application form and if it is published",grantSchemeId);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:71]:
```suggestion
        log.debug("Checking if scheme with id {} has internal application form and if it is published",grantSchemeId);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 7]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+
+        log.info("Getting Advert associated to scheme with id {}", grantSchemeId);
+        final GrantAdvert advert = grantAdvertService.getAdvertBySchemeId(grantSchemeId.toString());
+        log.info("Advert with id {} found for scheme with id {}", advert.getId(), grantSchemeId);
+
+        final String webpageUrl = grantAdvertService.getExternalSubmissionUrl(advert);
+
+        log.info("Checking that the advert has link to internal or external application");
+        final boolean webPageUrlIsForInternalApplications = webpageUrl.contains(environmentProperties.getFrontEndUri());
+        log.info("Advert is pointing to an internal application form : {}", webPageUrlIsForInternalApplications);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:85]:
```suggestion
        log.debug("Advert is pointing to an internal application form : {}", webPageUrlIsForInternalApplications);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 8]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+
+        final SchemeMandatoryQuestionApplicationFormInfosDto dto = SchemeMandatoryQuestionApplicationFormInfosDto.builder()
+                .hasInternalApplication(webPageUrlIsForInternalApplications)
+                .hasPublishedInternalApplication(false)
+                .build();
+
+        if (webPageUrlIsForInternalApplications) {
+            final boolean hasPublishedApplication = grantApplicationService.doesSchemeHaveAPublishedApplication(grantScheme);
+            log.info("Application form associated to advert with Id {} has PUBLISHED status: {}", advert.getId(), hasPublishedApplication);
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:94]:
```suggestion
            log.debug("Application form associated to advert with Id {} has PUBLISHED status: {}", advert.getId(), hasPublishedApplication);
```

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java [hunk 9]
```diff
@@ -53,4 +60,45 @@ public ResponseEntity<GetGrantSchemeWithApplicationAndAdverts> getGrantSchemeByI
...
+            dto.setHasPublishedInternalApplication(hasPublishedApplication);
+        }
+
+        log.info("Scheme with ID {} is for internal application: {} and has a published application form : {}",
```
Comment (ConnorTCO) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantSchemeController.java:98]:
```suggestion
        log.debug("Scheme with ID {} is for internal application: {} and has a published application form : {}",
```

**Review by dominicwest — APPROVED** [review]
Maybe some excessive logs but I'm not overly bothered by it!

**Pull request merged**